File: service/src/main/java/org/vsegda/request/MessageRequest.java
package org.vsegda.request;

import com.google.appengine.api.datastore.Entity;
import com.google.appengine.api.datastore.FetchOptions;
import com.google.appengine.api.datastore.Query;
import org.vsegda.data.MessageItem;
import org.vsegda.factory.DS;
import org.vsegda.storage.MessageStorage;
import org.vsegda.util.IdList;

import javax.servlet.http.HttpServletRequest;
import java.util.ArrayList;
import java.util.List;
import java.util.logging.Logger;

/**
 * @author Roman Elizarov
 */
public class MessageRequest extends AbstractRequest {
    private static final Logger log = Logger.getLogger(MessageRequest.class.getName());

	private IdList id;
    private boolean take;
	private long index;
    private int last = 100; // last 100 items by default
	private int first;

    public MessageRequest(HttpServletRequest req, boolean post) {
        init(req);
        if (id != null && id.isSingleton() && post)
            take = true; // force take on POST request if "id" is set
		if (index != 0 && (id == null || !id.isSingleton()))
            throw new IllegalArgumentException("cannot specify index without a singleton id");
        if (index != 0 && !take)
            throw new IllegalArgumentException("cannot specify index without take");
        if (take && (id == null || !id.isSingleton()))
            throw new IllegalArgumentException("cannot specify take without a singleton id");
        if (take && first != 0)
            throw new IllegalArgumentException("cannot specify take with first");
	}

    public List<MessageItem> query() {
        log.info("Performing message query " + this);
        long startTimeMillis = System.currentTimeMillis();
        List<MessageItem> items = new ArrayList<MessageItem>();
        if (id == null) {
            Query query = new Query("MessageQueue");
            query.addSort("__key__", Query.SortDirection.ASCENDING);
            Iterable<Entity> entities = DS.prepare(query).asIterable(
                    FetchOptions.Builder.withOffset(first).limit(last).chunkSize(last));
            for (Entity entity : entities) {
                items.add(MessageStorage.toMessageItem(entity));
            }
        } else {
            throw new UnsupportedOperationException(); // todo
//            for (String code : this.id) {
//                long id = MessageQueueService.resolveQueueCode(code);
//                Query query = DS.instance().newQuery(MessageItem.class);
//                if (take) {
//                    MessageQueue queue = DS.instance().getObjectById(MessageQueue.class, MessageQueue.createKey(id));
//                    index = Math.max(index, queue.getLastGetIndex());
//                    queue.setLastGetIndex(index);
//                    query.setFilter("queueId == id && messageIndex > index");
//                    query.declareParameters("long id, long index");
//                    query.setOrdering("messageIndex asc");
//                    query.setRange(first, first + last);
//                    items.addAll((Collection<MessageItem>)query.execute(id, index));
//                } else {
//                    query.setFilter("queueId == id");
//                    query.declareParameters("long id");
//                    query.setOrdering("messageIndex desc");
//                    query.setRange(first, first + last);
//                    items.addAll((Collection<MessageItem>)query.execute(id));
//                }
//            }
        }
        log.info("Completed message query in " + (System.currentTimeMillis() - startTimeMillis) + " ms");
        return items;
    }

    public IdList getId() {
        return id;
    }

    public void setId(IdList id) {
        this.id = id;
    }

    public boolean isTake() {
        return take;
    }

    public void setTake(boolean take) {
        this.take = take;
    }

    public long getIndex() {
        return index;
    }

    public void setIndex(long index) {
        this.index = index;
    }

    public int getFirst() {
        return first;
    }

    public void setFirst(int first) {
        this.first = first;
    }

    public int getLast() {
        return last;
    }

    public void setLast(int last) {
        this.last = last;
    }
}

File: service/src/main/java/org/vsegda/request/MessageRequest.kt
package org.vsegda.request

import io.ktor.request.*
import org.vsegda.data.*
import org.vsegda.service.*
import org.vsegda.util.*
import javax.servlet.http.*

class MessageRequest() : AbstractRequest() {
    var id: IdList? = null
    var isTake: Boolean = false
    var index: Long = 0
    var last = 100 // last 100 items by default
    var first: Int = 0

    constructor(req: HttpServletRequest, post: Boolean) : this() {
        init(req)
        validate(post)
    }

    constructor(req: ApplicationRequest, post: Boolean = false) : this() {
        init(req)
        validate(post)
    }

    private fun validate(post: Boolean) {
        if (id != null && id!!.isSingleton && post)
            isTake = true // force take on POST request if "id" is set
        if (index != 0L && (id == null || !id!!.isSingleton))
            throw IllegalArgumentException("cannot specify index without a singleton id")
        if (index != 0L && !isTake)
            throw IllegalArgumentException("cannot specify index without take")
        if (isTake && (id == null || !id!!.isSingleton))
            throw IllegalArgumentException("cannot specify take without a singleton id")
        if (isTake && first != 0)
            throw IllegalArgumentException("cannot specify take with first")
    }

    fun query(): List<MessageItem> =
        logged("Message request $this", around = true, result = { "${it.size} items" }) {
            val id = this.id
            if (id == null) {
                MessageQueueService.messageQueues
                    .map {
                        MessageItemService.getMessageItem(it.queueId, it.lastPostIndex) ?:
                            MessageItem(it.queueId, it.lastPostIndex)
                    }
            } else {
                id.asSequence()
                    .mapNotNull { MessageQueueService.resolveMessageQueueByCode(it) }
                    .flatMap { queue ->
                        val list = if (isTake) {
                            index = Math.max(index, queue.lastGetIndex)
                            queue.lastGetIndex = index
                            MessageQueueService.updateMessageQueue(queue)
                            MessageItemService.getNewMessageItems(queue.queueId, index, first, last)
                        } else {
                            MessageItemService.getLastMessagesItems(queue.queueId, first, last)
                        }
                        list.asSequence()
                    }
                    .toList()
            }
        }
}